platform/engineering/design-docs/2020-08-17-mobile-api-design-doc.md

```diff
@@ -0,0 +1,173 @@
+integrations. This will allow the mobile API to grow independently of the rest of vets-api, while providing a consistent API 
+that will be useful for future mobile applications to consume.
+
+All endpoints will be versioned, so that long-lasting mobile releases will continue to function into the future. When 
+modifications are required, changes will be made in a way to not break existing clients. The team is considering 
+building a machine-readable discovery API endpoint to decouple the mobile app from hard-coded urls.
+
+All mobile authorization will be confined to this Rails engine. The mobile app will require some new endpoints as well, and 
+these will be built in the Rails engine for isolation. [A companion auth design doc has more details](https://github.com/department-of-veterans-affairs/va.gov-team/pull/12482).
+
+## Specifics
+
+### Detailed Design
+The mobile endpoints need custom authentication, merged responses, and reduced payloads. With versioning this is possible 
+in the core api albeit with a fractured experience. Mobile endpoints could be on v0, v1, or even v2. MVP mobile endpoints 
+will be in flux, and parallel va.gov changes could cause versions to climb.
+
+To limit the disruption to the API it's best to perform the optimizations in a module with a `/mobile` base path. 
+This allows us to:
+
+- Add mobile authentication without changes to the core API codebase.
+- Stay out of the way as we tweak endpoints in a mobile sandbox.
+- Test out optimizations that can be later ported back to the core API.
+
+Expanding on the last point it's possible that once the mobile optimized versions of the endpoints have stabilized that we permanently
+move them back to core API. The aim is to avoid confusion about why a pereceived newer version of an endpoint would be missing fields
+without context that it was actually a 'mobile only' endpoint.
+
+#### Application Controller
+
+Authentication for the mobile app is via IAM's SSOe OAuth service. It is 100% API based and does not use callbacks or cookies. 
+The mobile `ApplicationController` would extend the core `ApplicationController` and override its authentication methods.
+
+Enabling the mobile API will expose some tech debt in the `ApplicationController` that is not a requirement to fix; but it would clean up things for all its sub-classes.
+The logic in the core `ApplicationController` that does not apply globally could move to concerns be and 
+included as needed. e.g. `render_job_id` only applies to the `DocumentsController` and the `EVSSClaimsController`. 
+
+#### Routing
+
+The mobile engine's endpoints mount at `/mobile`:
+
+```ruby
+mount Mobile::Engine, at: '/mobile'
+```
+
+Some endpoints such as facilities may not require optimization. They will pass-through via routes:
+
+```ruby
+# modules/mobile/config/routes.rb (redirect starts at the app root so this is /v0/mobile/.. pointing to )
+get '/v0/facilities/va/:id' => redirect('/v1/facilities/va/%{id}')
```

Comment:
to clarify, this would indicate that calls to `api.va.gov/v0/mobile/v0/facilities/va/:id` would auto-route to `api.va.gov/v1/facilities/va/:id`?

assuming yes, could you could talk about what kind of testing would ensure that a change to the `/v1/facilities/va/:id` endpoint output doesn't break mobile? (or rather, at least alerting developers that the proposed change affects mobile apps, not just website clients)

Reply:
I think the `redirect` here returns a 301 to the client, which would then result in the mobile client making a subsequent request directly to `/v1/facilities/va/%{id}`, bypassing the `mobile` module endpoint(s). I'm not sure if that's the desired behavior.  Please correct me if I'm wrong.

Reply:
Yes @jholton is correct it is 301 which is not how we want the route to work. Instead we'll have our own route definitions, controllers, and serializers. There would be some code duplication. Specs would guard services from changing out from under us.